### PR TITLE
cgc7900: attach keyboard and serial port, force cold boot

### PR DESCRIPTION
**Pull request by shattered**

### CODE REVIEW

File: src/mame/drivers/cgc7900.cpp [hunk 1]
```diff
@@ -128,40 +170,99 @@ WRITE16_MEMBER( cgc7900_state::keyboard_w )
...
+			}
+		}
+	}
+
 	m_int_mask = data;
 }
 
+void cgc7900_state::irq_encoder(int pin, int state)
+{
+	if(state == ASSERT_LINE)
```
Comment (cuavas) [inline src/mame/drivers/cgc7900.cpp]: Consistency - this one has no space after, the others do.

File: src/mame/drivers/cgc7900.cpp [hunk 2]
```diff
@@ -128,40 +170,99 @@ WRITE16_MEMBER( cgc7900_state::keyboard_w )
...
+WRITE_LINE_MEMBER(cgc7900_state::irq_6) { irq_encoder(6, state); }
+WRITE_LINE_MEMBER(cgc7900_state::irq_7) { irq_encoder(7, state); }
+WRITE_LINE_MEMBER(cgc7900_state::irq_8) { irq_encoder(8, state); }
+WRITE_LINE_MEMBER(cgc7900_state::irq_9) { irq_encoder(9, state); }
+WRITE_LINE_MEMBER(cgc7900_state::irq_a) { irq_encoder(0xa, state); }
+WRITE_LINE_MEMBER(cgc7900_state::irq_b) { irq_encoder(0xb, state); }
+WRITE_LINE_MEMBER(cgc7900_state::irq_c) { irq_encoder(0xc, state); }
+WRITE_LINE_MEMBER(cgc7900_state::irq_d) { irq_encoder(0xd, state); }
+WRITE_LINE_MEMBER(cgc7900_state::irq_e) { irq_encoder(0xe, state); }
+WRITE_LINE_MEMBER(cgc7900_state::irq_f) { irq_encoder(0xf, state); }
```
Comment (cuavas) [inline src/mame/drivers/cgc7900.cpp]: Please use a template for this to reduce copy-paste, i.e. in the header at declaration point:
```
template <unsigned N> DECLARE_WRITE_LINE_MEMBER(irq) { irq_encoder(N, state); }
```

Then you can get each of them as `cgc7900_state::irq<0x0>`, `cgc7900_state::irq<0x1>`, ..., `cgc7900_state::irq<0xf>` and not need 16 of the things here.

File: src/mame/includes/cgc7900.h [hunk 3]
```diff
@@ -6,12 +6,18 @@
 #define __CGC7900__
```
Comment (cuavas) [inline src/mame/includes/cgc7900.h]: While you're changing this file, could you change the `#include` guard to follow the convention avoiding reserved `__NAMES`?  It should be `MAME_INCLUDES_CGC7900_H`.

File: src/mame/includes/cgc7900.h [hunk 4]
```diff
@@ -81,19 +95,50 @@ class cgc7900_state : public driver_device
...
 	/* video state */
 	rgb_t m_clut[256];
 	int m_blink;
 
 	TIMER_DEVICE_CALLBACK_MEMBER(blink_tick);
+
+	void kbd_put(u8 data);
+
+private:
+	u16 kbd_mods;
```
Comment (cuavas) [inline src/mame/includes/cgc7900.h]: Is there any reason you've only made this stuff private and not a whole lot of other stuff as well?  There's no reason for all the device members to be public.

Comment (shattered) [reply src/mame/includes/cgc7900.h]: It's used by MCFG_GENERIC_KEYBOARD_CB macro and has to be public.

Comment (cuavas) [reply src/mame/includes/cgc7900.h]: I was referring to all the `required_device` and `required_shared_ptr` stuff above, and the overrides of protected member functions.  I realise the keyboard put callback needs to be public.  I know it was like that already, but it's nice to do "drive-by cleanup" when doing stuff like this.

**Comment by cuavas** [comment]
Merging for now.  If you do further work on this, could you clean up the public members?

**Pull request merged**